### PR TITLE
Tensor Parallelism Training and Inference

### PR DESCRIPTION
**Pull request by vince62s**
Goal is to use torch.distributed and split Modules when possible (along Columns or Rows) so that bigger models can be split across multiple GPU.
New option: parallel_mode = "data_parallel" by default, and "tensor_parallel" to activate tensor parallelism.

### CODE REVIEW

File: onmt/models/model.py [hunk 1]
```diff
@@ -148,9 +151,39 @@ def load_safe_state_dict(
...
-                            name + "." + param_name
-                        )
+
+                        if name.split(".")[-1] in [
+                            "linear_keys",
+                            "linear_values",
+                            "linear_query",
+                            "w_1",
+                            "w_3",
+                        ]:
```
Comment (francoishernandez) [inline onmt/models/model.py:161]: do we want to make this configurable (like quant_layers)?

File: onmt/modules/multi_headed_attn.py [hunk 2]
```diff
@@ -246,45 +247,53 @@ def __init__(
...
         super(MultiHeadedAttention, self).__init__()
         self.head_count = head_count
         self.num_kv = num_kv
+        self.parallel_gpu = parallel_gpu
         if num_kv == 0:
             self.linear_keys = skip_init(
                 nn.Linear,
                 in_features=model_dim,
-                out_features=model_dim,
+                out_features=model_dim // parallel_gpu,
```
Comment (francoishernandez) [inline onmt/modules/multi_headed_attn.py:262]: gentle reminder to put some assert somewhere for edge cases where model_dim might not be divisible by parallel_gpu (valid for other divisions below)

File: onmt/utils/distributed.py [hunk 3]
```diff
@@ -176,3 +181,110 @@ def consumer(process_fn, opt, device_id, error_queue):  # noqa: E501
...
+        pass  # killed by parent, do nothing
+    except Exception:
+        # propagate exception to parent process, keeping original traceback
+        import traceback
+
+        error_queue.put((opt.gpu_ranks[device_id], traceback.format_exc()))
+
+
+class mp_inference(object):
+    def __init__(self, opt):
```
Comment (francoishernandez) [inline onmt/utils/distributed.py:240]: Maybe add some docstrings. Also weird to have the classname in snake_case. `MPInference` instead?

File: eval_llm/MMLU/run_mmlu_opennmt.py [hunk 4]
```diff
@@ -180,30 +175,23 @@ def evaluate(opt):
...
-            align_debug=opt.align_debug,
-        )
         pred_answers = [
             x.lstrip() for sublist in preds for x in sublist
         ]  # flatten the list of list
 
         gold_answers = [record["answer"] for record in records]
         run_results[task] = {"pred_answers": pred_answers, "gold_answers": gold_answers}
+
+    engine.terminate()
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eval_llm/MMLU/run_mmlu_opennmt.py:187]: indeed. But also we may ask ourselves if we can make this the mainstream way, for translate.py for instance.

File: onmt/model_builder.py [hunk 5]
```diff
@@ -116,22 +127,29 @@ def load_test_model(opt, model_path=None):
...
         # weights are in the .pt file
         model.load_state_dict(
-            checkpoint, precision=precision, device=device, strict=True
+            checkpoint,
+            precision=precision,
+            device=device,
+            strict=True,
+            # world_size=model_opt.world_size,
+            # parallel_mode=model_opt.parallel_mode,
+            # device_id=device_id,
```
[the remark this answers is not in the document]
Comment (vince62s) [reply onmt/model_builder.py:140]: No, I will implement it quickly.

File: onmt/modules/multi_headed_attn.py (hunk 2 above)
Comment (vince62s) [reply onmt/modules/multi_headed_attn.py:262]: will do.

File: eval_llm/MMLU/run_mmlu_opennmt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vince62s) [reply eval_llm/MMLU/run_mmlu_opennmt.py:187]: Second thought will do it later in another PR when trying to document a "Library / API - like" page.

File: eval_llm/MMLU/run_mmlu_opennmt.py [hunk 6]
```diff
@@ -7,13 +7,14 @@
 import time
 import pandas as pd
 from onmt.utils.logging import init_logger
-from onmt.utils.distributed import MPInference
+from onmt.inference_engine import InferenceEngine
```
Comment (francoishernandez) [inline eval_llm/MMLU/run_mmlu_opennmt.py:10]: I think we miss the new `inference_engine` file in this commit :upside_down_face:

**Pull request merged**